components/sup/src/manager.rs

```diff
@@ -1267,36 +1247,55 @@ impl Manager {
             .butterfly_data = json;
     }
 
-    fn persist_services_state(&self) {
+    fn persist_services_state(&mut self, runtime: &mut Runtime) {
         let config_rendering = if self.feature_flags.contains(FeatureFlag::REDACT_HTTP) {
             ConfigRendering::Redacted
         } else {
             ConfigRendering::Full
         };
 
-        let services = self.state
-                           .services
-                           .read()
-                           .expect("Services lock is poisoned!");
-        let existing_idents: Vec<PackageIdent> =
-            services.values().map(|s| s.spec_ident.clone()).collect();
-
+        let existing_idents: Vec<PackageIdent> = self.state
+                                                     .services
+                                                     .read()
+                                                     .expect("Services lock is poisoned!")
+                                                     .values()
+                                                     .map(|s| s.spec_ident.clone())
+                                                     .collect();
         // Services that are not active but are being watched for changes
         // These would include stopped persistent services or other
         // persistent services that failed to load
         let watched_services: Vec<Service> =
             self.spec_dir
                 .specs()
                 .iter()
-                .filter(|spec| !existing_idents.contains(&spec.ident))
-                .flat_map(|spec| {
-                    Service::load(self.sys.clone(),
-                                  spec.clone(),
-                                  self.fs_cfg.clone(),
-                                  self.organization.as_ref().map(|org| &**org),
-                                  self.state.gateway_state.clone()).into_iter()
+                .filter_map(|spec| {
+                    if existing_idents.contains(&spec.ident) {
+                        None
+                    } else {
+                        match Service::load(self.sys.clone(),
+                                            spec.clone(),
+                                            self.fs_cfg.clone(),
+                                            self.organization.as_ref().map(|org| &**org),
+                                            self.state.gateway_state.clone())
+                        {
+                            Ok(s) => Some(s),
+                            Err(e) => {
+                                outputln!("Failed to load spec file for '{}' with error '{}'. \
+                                           This service will be unloaded.",
+                                          spec.ident,
+                                          e);
+                                self.unload(spec, &ShutdownInput::default(), runtime);
```

Comment:
This refactoring highlights one of the things I've always felt was a little "weird" about this function, as well as how our other functions around Services are named. Perhaps now is the time to address it.

This function is just about persisting our state for the HTTP gateway. Thus, one should expect that it wouldn't change the broader state of the Supervisor. The existing state of the Supervisor should be "read-only" for this function - it should only be changing the state of the HTTP gateway (but really, we should have a function that just reads Supervisor state and creates the new gateway state, which we then pass to another function that modifies the gateway itself, but I digress). Having to pass a Tokio Runtime into this function is thus _super_ suprising and unexpected.

Then we come to the issue of naming and how existing functions are being used, with a side-trip into "ubiquitous language". From a user's perspective, "load" means "make the Supervisor aware of a given service", making it possible to run that service; "unload" means the opposite. However, once we get into the code, `Service::load` is really a constructor... I think it should be `Service::new`, really. The "load" operation is really something that takes place well before we get into the `Service` code, since "loading" is "create a spec file on disk". That we can even create a `Service` instance implies that that information has _already_ been loaded into the Supervisor.

Thus, if `Service::load` is really a constructor, then this `unload` operation isn't symmetric, because it is actually stopping a service as well as removing a spec file from disk.

Stepping back a bit, it seems that the only reason we are going through this "load" logic (very intentionally using air-quotes here), is because we don't have an in-memory representation of services that are loaded, but not running (e.g., because they are supposed to be stopped). If we had _that_, then we wouldn't need to constantly construct these `Service` instances. The persistence call could then just simply read Supervisor state. Additionally, we could push the handling of unloadable services out toward the periphery, and handle them when we first encounter them, rather than deeper in the code like this.